apps/grader-host/src/index.ts

```diff
@@ -188,24 +207,80 @@ async function handleJob(job: GradingJobMessage) {
   try {
     const receivedTime = await timeReporter.reportReceivedTime(job.jobId);
 
-    const initResults = await Promise.all([
+    const initResults = await Promise.allSettled([
```

Comment:
Why do we want to run all of them to completion? We want to fail if any of them fails and I don't think we care about exactly how many fail in that case?

Reply:
I want to be able to tell _what_ has actually failed. A `Promise.all(...)` wrapped in a `try`/`catch` doesn't let me do that; `allSettled` does.

Reply:
Understood. I was thinking about handling the "what" with the original calls, to avoid needing to do all this stuff with `initResults[2]` at all. Something like:
```
const messages = [];
let tempDir;
await Promise.all([
  reportReceived(context, receivedTime).catch(() => messages.push('Could not report received.')),
  initDocker(context).catch(() => messages.push(`Could not pull Docker image ${job.image}.`)),
  initFiles(context).then(
    (tempDirResult) => {
      tempDir = tempDirResult;
    },
    (err) => {
      Sentry.captureException(err);
      throw err;
    }
  ),
]);

const result = messages.length == 0
  ? await runJob(context, receivedTime, tempDir?.tempDir)
  : {
    job_id: job.jobId,
    received_time: receivedTime,
    start_time: new Date(),
    end_time: new Date(),
    succeeded: false,
    message: messages.join('\n'),
  };
```

But this is all just bike-shedding, your code works fine so go with whatever style you prefer here.

Reply:
Ah, gotcha! Yeah, I'm going to stick with what I have here.